### PR TITLE
Improve CosmosVFS api

### PR DESCRIPTION
**Pull request by MishaTY**
Improves the CosmosVFS api. Also adds support for creating MBR partitions

### CODE REVIEW

**Comment by KM198912** [comment]
Can you remove the iss file from the PR?

**Comment by terminal-cs** [comment]
when merge

**Comment by ThomasBeHappy** [comment]
Does this also fix the 1mb problem we got?

**Comment by MishaTY** [comment]
Nope, but I will fix that in a future PR.

**Review by valentinbreiz — DISMISSED** [review]

**Comment by valentinbreiz** [comment]
@MishaTY would it be possible to add a test for creating new MBR partitions?

File: Tests/Cosmos.System.Tests/DiskManagerTest.cs [hunk 1]
```diff
@@ -36,24 +56,27 @@ public void CleanUp()
...
         [Test]
-        public void TestDiskManagerQuickFormat()
+        public void Execute()
         {
-            string driveName = @"C:\";
-            var drive = new DiskManager(driveName);
-            drive.Format("FAT32", aQuick: true);
+            string driveName = @"0:\";
+
+            Assert.IsTrue(ourPart.RootPath == driveName, "ManagedPartition.RootPath failed drive has wrong name");
```
Comment (quajak) [inline Tests/Cosmos.System.Tests/DiskManagerTest.cs:63]: Nitpick: Use Assert.AreEqual not IsTrue in future, it provides more helpful error messages when tests fail.

File: source/Cosmos.HAL2/BlockDevice/IDE.cs [hunk 2]
```diff
@@ -49,68 +53,55 @@ private static void Initialize(Ata.ControllerIdEnum aControllerID, Ata.BusPositi
...
-                        Console.WriteLine("Null partition found at idx: " + i);
-                    }
-                    else
-                    {
-                        var xPartDevice = new Partition(xATA, xPart.StartSector, xPart.SectorCount);
-                        BlockDevice.Devices.Add(xPartDevice);
-                        Console.WriteLine("Found partition at idx: " + i);
-                    }
+                    Partition.Partitions.Add(new Partition(device, part.StartSector, part.SectorCount));
+                    Console.WriteLine("Found partition at idx: " + i);
```
Comment (quajak) [inline source/Cosmos.HAL2/BlockDevice/IDE.cs:77]: Do we need to write this to the screen?

File: source/Cosmos.HAL2/BlockDevice/IDE.cs [hunk 3]
```diff
@@ -49,68 +53,55 @@ private static void Initialize(Ata.ControllerIdEnum aControllerID, Ata.BusPositi
...
-                    }
-                    else
-                    {
-                        var xPartDevice = new Partition(xATA, part.StartSector, part.SectorCount);
-                        BlockDevice.Devices.Add(xPartDevice);
-                        Console.WriteLine("Found partition at idx: " + partNumb);
-                    }
-                    partNumb++;
+                    Partition.Partitions.Add(new Partition(device, part.StartSector, part.SectorCount));
+                    Console.WriteLine("Found partition at idx: " + c);
```
Comment (quajak) [inline source/Cosmos.HAL2/BlockDevice/IDE.cs:103]: Same here.

File: source/Cosmos.System2/FileSystem/CosmosVFS.cs [hunk 4]
```diff
@@ -181,7 +185,7 @@ public override DirectoryEntry CreateDirectory(string aPath)
 
             if (aPath == null)
             {
-                throw new ArgumentNullException(nameof(aPath));
+                throw new ArgumentNullException("path is: "+nameof(aPath));
```
Comment (quajak) [inline source/Cosmos.System2/FileSystem/CosmosVFS.cs:188]: Why change the error message?

File: source/Cosmos.System2/FileSystem/Disk.cs [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+        private static List<FileSystemFactory> registeredFileSystems = new List<FileSystemFactory>();
+        public bool IsMBR { get { return !GPT.IsGPTPartition(Host); } }
+        /// <summary>
+        /// The size of the disk in MB.
+        /// </summary>
+        public int Size { get; }
+        /// <summary>
+        /// List of partitions
+        /// </summary>
+        public List<ManagedPartition> Partitions
```
Comment (quajak) [inline source/Cosmos.System2/FileSystem/Disk.cs:23]: Does it make sense to cache the partitions after they have been found once?

File: source/Cosmos.System2/FileSystem/Disk.cs [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+        public static List<FileSystemFactory> RegisteredFileSystemsTypes { get { return registeredFileSystems; } }
+        /// <summary>
+        /// Main blockdevice that has all of the partitions.
+        /// </summary>
+        public BlockDevice Host;
+        public BlockDeviceType Type { get { return Host.Type; } }
+        public Disk(BlockDevice mainBlockDevice)
+        {
+            Host = mainBlockDevice;
+            if (registeredFileSystems.Count == 0)
```
Comment (quajak) [inline source/Cosmos.System2/FileSystem/Disk.cs:78]: CAn we not set a default value for the List including the FatFileSystemFactory already?

File: source/Cosmos.System2/FileSystem/Disk.cs [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+{
+    public class Disk
+    {
+        private List<ManagedPartition> parts = new List<ManagedPartition>();
+        private static List<FileSystemFactory> registeredFileSystems = new List<FileSystemFactory>();
+        public bool IsMBR { get { return !GPT.IsGPTPartition(Host); } }
+        /// <summary>
+        /// The size of the disk in MB.
+        /// </summary>
+        public int Size { get; }
```
Comment (quajak) [inline source/Cosmos.System2/FileSystem/Disk.cs:19]: Why in MB?

File: source/Cosmos.System2/FileSystem/Disk.cs [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+                FatFileSystem.CreateFatFileSystem(part.Host, VFSManager.GetNextFilesystemLetter() + ":\\", xSize, format);
+                Mount();
+            }
+            else
+            {
+                throw new NotImplementedException(format + " formatting not supported.");
+            }
+        }
+
+        private FileSystem[] MountedPartitions = new FileSystem[4];
```
Comment (quajak) [inline source/Cosmos.System2/FileSystem/Disk.cs:258]: Why does it have length 4?

File: source/Cosmos.System2/FileSystem/ManagedPartition.cs [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+        public ManagedPartition(Partition host)
+        {
+            Host = host;
+        }
+        /// <summary>
+        /// Zeros out the entire partition
+        /// </summary>
+        public void Clear()
+        {
+            for (ulong i = 0; i < Host.BlockCount; i++)
```
Comment (quajak) [inline source/Cosmos.System2/FileSystem/ManagedPartition.cs:44]: Should this also set `MountedFS = null`?

**Review by quajak — CHANGES_REQUESTED** [review]
General approach looks good. Just a few minor points. It would be cool if you could add a short article explaining the general approach so its easier for new developers and users to understand.

File: source/Cosmos.System2/FileSystem/Disk.cs (hunk 5 above)
Comment (MishaTY) [reply source/Cosmos.System2/FileSystem/Disk.cs:23]: Not really, as the partitions can change when Add/Remove partitions is called.

File: source/Cosmos.System2/FileSystem/Disk.cs (hunk 8 above)
Comment (MishaTY) [reply source/Cosmos.System2/FileSystem/Disk.cs:258]: Because there can only be 4 partitions in mbr partitioning scheme.

Comment (quajak) [reply source/Cosmos.System2/FileSystem/Disk.cs:262]: Ok thanks

**Review by valentinbreiz — APPROVED** [review]

**Review by quajak — APPROVED** [review]

**Pull request merged**